test/e2e/networking_perf.go

```diff
@@ -0,0 +1,249 @@
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/intstr"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	// This is a magic number, don't rely on it as a well defined constant.
+	gceBandwidthBitsEstimate = int64(30000000000)
+)
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
+var _ = KubeDescribe("Networking IPerf [Experimental] [Feature:Network-Performance] ", func() {
+
+	f := NewDefaultFramework("network-perf")
+
+	// A few simple bandwidth tests which are capped by nodes.
+	runClientServerBandwidthMeasurement(f, 1, gceBandwidthBitsEstimate)
+	runClientServerBandwidthMeasurement(f, 5, gceBandwidthBitsEstimate)
+	runClientServerBandwidthMeasurement(f, 8, gceBandwidthBitsEstimate)
+
+	// As many clients as possible based on node count.
+	runClientServerBandwidthMeasurement(f, 10000, gceBandwidthBitsEstimate)
+})
+
+func runClientServerBandwidthMeasurement(f *Framework, numClient int, maxBandwidthBits int64) {
+
+	// TODO: Make this a function parameter, once we distribute iperf endpoints.
+	numServer := 1
+
+	It(fmt.Sprintf("should transfer ~ 1GB onto the service endpoint %v servers (maximum of %v clients)", numServer, numClient), func() {
+		nodes := ListSchedulableNodesOrDie(f.Client)
+		totalPods := len(nodes.Items)
+		// for a single service, we expect to divide bandwidth between the network.  Very crude estimate.
+		expectedBandwidth := int(float64(maxBandwidthBits) / float64(totalPods))
+		Expect(totalPods).NotTo(Equal(0))
+		appName := "iperf-e2e"
+		CreateServiceWithPods(f,
+			8001,
+			8002,
+			appName,
+			api.Container{
+				Name:  "iperf-server",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-s", "-p", "8001"},
+				Ports: []api.ContainerPort{{ContainerPort: 8001}},
+			},
+			// this will be used to generate the -service name which all iperf clients point at.
+			numServer, // 1 server
+		)
+		CreateServiceWithPods(f,
+			-1,
+			-1,
+			"iperf-e2e-cli", // this will be used to generate the -service name which all iperf clients point at.
+			api.Container{
+				Name:  "iperf-client",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-c", "service-for-" + appName, "-p", "8002", "--reportstyle", "C",
+				},
+			},
+			numClient, // 1 server
+		)
+		Logf("Reading all perf results to stdout.")
+		Logf("date,cli,cliPort,server,serverPort,id,interval,transferBits,bandwidthBits")
+		bandwidthMap := make(map[string]int64)
+
+		// Calculate expected number of clients based on total nodes.  ITs own function since this
+		// logic may change over time at scale.
+		expectedCli := func() int {
+			nodes, err := GetReadyNodes(f)
+			expectNoError(err)
+			return int(math.Min(float64(len(nodes.Items)), float64(numClient)))
+		}()
+
+		start := time.Now()
+		for len(bandwidthMap) < expectedCli && time.Since(start) <= 2*time.Minute {
```

Comment:
Hmmm... Not sure if this should be a constant.  I think really we should scale it w/respect to client number.... but point taken - it should either be a constant or it should be dynamic./